extensions/spark/kyuubi-spark-jvm-quake/src/main/scala/org/apache/spark/kyuubi/jvm/quake/KyuubiJVMQuake.scala

```diff
@@ -0,0 +1,147 @@
+import org.apache.spark.internal.Logging
+import org.apache.spark.kyuubi.jvm.quake.KyuubiConf._
+import org.apache.spark.util.ThreadUtils
+
+class KyuubiJVMQuake(conf: SparkConf, heapDumpEnabled: Boolean) extends Logging {
+  private val appId = conf.get("spark.app.id", "app-id").stripPrefix("application_")
+  private val dumpThreshold = TimeUnit.SECONDS.toNanos(conf.get(JVM_QUAKE_DUMP_THRESHOLD))
+  private val killThreshold = TimeUnit.SECONDS.toNanos(conf.get(JVM_QUAKE_KILL_THRESHOLD))
+  private val exitCode = conf.get(JVM_QUAKE_EXIT_CODE)
+  private val checkInterval = TimeUnit.SECONDS.toMillis(conf.get(JVM_QUAKE_CHECK_INTERVAL))
+  private val runTimeWeight = conf.get(JVM_QUAKE_RUN_TIME_WEIGHT)
+  private[quake] val heapPath = s"${conf.get(JVM_QUAKE_HEAP_DUMP_PATH)}/$appId"
+
+  private[quake] var (lastExitTime, lastGCTime) = getLastGCInfo
+  private[quake] var bucket: Long = 0L
+  private[quake] var heapExist: Boolean = false
+
+  private val scheduler: ScheduledExecutorService =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("jvm-quake")
+
+  private[quake] def run(): Unit = {
+    val (currentExitTime, currentGcTime) = getLastGCInfo
+    if (currentExitTime == lastExitTime) {
+      return
+    }
+    val gcTime = currentGcTime - lastGCTime
+    val runTime = currentExitTime - lastExitTime - gcTime
+
+    bucket = Math.max(0, bucket + gcTime - BigDecimal(runTime * runTimeWeight).toLong)
+
+    if (bucket > dumpThreshold) {
+      logError(s"JVM GC has reached the threshold!!!" +
+        s" (bucket: ${bucket / 1000000000}s, dumpThreshold: ${dumpThreshold / 1000000000}s)")
+      if (shouldDumpHeap) {
+        saveHeap(heapPath)
+        heapExist = true
+      } else if (bucket > killThreshold) {
+        System.exit(exitCode)
+      }
+    }
+
+    lastExitTime = currentExitTime
+    lastGCTime = currentGcTime
+  }
+
+  def start(): Unit = {
+    scheduler.scheduleAtFixedRate(
+      new Runnable() {
+        override def run(): Unit = {
+          KyuubiJVMQuake.this.run()
+        }
+      },
+      0,
+      checkInterval,
+      TimeUnit.MILLISECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+  }
+
+  private def shouldDumpHeap: Boolean = {
+    heapDumpEnabled && !heapExist
+  }
+
+  private def saveHeap(savePath: String): Unit = {
+    try {
+      val saveDir = new File(savePath)
+      if (!saveDir.exists()) {
+        saveDir.mkdirs()
+      }
+      val heapDumpFile = new File(saveDir, s"kyuubi-jvm-quake-heapdump.hprof")
+      if (heapDumpFile.exists()) {
+        logInfo(s"Heap exits $heapDumpFile")
+        heapExist = true
+        return
+      }
+      logInfo(s"Starting heap dump at $heapDumpFile")
+      val server = ManagementFactory.getPlatformMBeanServer
+      val mxBean = ManagementFactory.newPlatformMXBeanProxy(
+        server,
+        "com.sun.management:type=HotSpotDiagnostic",
+        classOf[HotSpotDiagnosticMXBean])
+      mxBean.dumpHeap(heapDumpFile.getAbsolutePath, false)
```

Comment:
btw, are these apis currently stable between different jdk versions?

Reply:
Tested different types of JDKs and they all worked fine